### ISSUE TITLE
Reciprocal learning questions

### ISSUE THREAD

**Issue by tabacof** [facebookresearch/kbc#12]
Hi, first of all many thanks for releasing the code alongside the paper. 

On table 2 of the paper we see results with and without reciprocal learning. By taking a look and playing with the code, I realised reciprocal learning seems to be hardcoded to the way the datasets are loaded and the embeddings are defined.

I have two questions then:

1. Is it possible to run the code without reciprocal learning to fully reproduce the results of table 2?
2. Is the reciprocal evaluation on the right-hand side only totally comparable to other approaches where there are corruptions to both the subject and the object?

Thanks again.

**Comment by timlacroix** [comment]
Hi, the code currently on the repo is a simplification of the original code. It is meant to make understanding the method presented in the paper and reproducing the results with reciprocals included easier (since this is the method that is presented in the paper).

If you want to reproduce the results of the paper outside of this setting, you should be able to do so with the cpp version of the code which you can find at this commit I think : https://github.com/facebookresearch/kbc/commit/8cb8539001df2c6868179149cd81a8f66729371d

I'm not sure I understand the second question could you re-formulate ? 

thanks,

Tim'

**Comment by tabacof** [comment]
Hi @timlacroix, thanks for the quick response.

So with respect two (2), let me try expand a little more: The standard evaluation protocol for missing link prediction is applying corruption to a test set and then ranking. 

We get all the entities and corrupt both the subject and object of one particular test triple with them, and then see how well ranked is the true triple among all the corruptions, excluding other known true triples (e.g. from the training set) from the ranking evaluation.

Do you agree with my formulation? If so, would you say the code implements that same procedure? I am having trouble getting my head around how the code does that.

**Comment by timlacroix** [comment]
"We get all the entities and corrupt both the subject and object of one particular test triple with them, and then see how well ranked is the true triple among all the corruptions, excluding other known true triples (e.g. from the training set) from the ranking evaluation."
-> Not really, here is how I would formulate the evaluation: We take one true triple (head, predicate, tail), and rank either head or tail, among all other entities in the knowledge base, ignoring true triples. I would say this is what is implemented in my paper and what is standard in the literature.

Regarding how this is implemented :

Starting in datasets.py 
https://github.com/facebookresearch/kbc/blob/master/kbc/datasets.py#L50

Either we evaluate with missing lhs or with missing rhs. Since the method we call in models.py assumes missing rhs, we have to reverse the triples and change the relation ids (to make them reciprocals) in case of missing lhs (here https://github.com/facebookresearch/kbc/blob/master/kbc/datasets.py#L68).

Then in models.py https://github.com/facebookresearch/kbc/blob/master/kbc/models.py#L27
We start by getting all possible rhs embeddings (by calling the get_rhs method on the model). And create the queries corresponding to tuples (lhs, rel) for a single batch https://github.com/facebookresearch/kbc/blob/master/kbc/models.py#L49.

Once this is done, all that remains is to evaluate the filtered rank. We compute all scores by doing a matrix matrix product. Then, for all true triples to filter out, we set their scores to -1e6 (effectively making sure that they are ranked last). On line 68, we compute the number of scores that are higher or equal to the score of the triple we are ranking (in targets), this is exactly the filtered rank.


Is this clearer ?

**Comment by tabacof** [comment]
> Not really, here is how I would formulate the evaluation: We take one true triple (head, predicate, tail), and rank either head or tail, among all other entities in the knowledge base, ignoring true triples. I would say this is what is implemented in my paper and what is standard in the literature.

Sorry, that is what I actually intended to express.

> Is this clearer ?

Yes, definitely, thank you very much for the explanation.
